netcanv-renderer-opengl/src/font.rs

```diff
@@ -204,28 +229,36 @@ impl netcanv_renderer::Font for Font {
    }
 }
 
-pub(crate) struct GlyphRenderer<'face, 'store, 'gl> {
-   face: &'face Face,
+pub(crate) struct GlyphRenderer<'font, 'store, 'gl, 'c> {
+   swash_font: FontRef<'font>,
    size_store: &'store mut FontSize,
    gl: &'gl glow::Context,
+   scale_context: &'c mut ScaleContext,
+   shape_context: &'c mut ShapeContext,
 }
 
-impl<'font, 'store, 'gl> GlyphRenderer<'font, 'store, 'gl> {
+impl<'font, 'store, 'gl, 'c> GlyphRenderer<'font, 'store, 'gl, 'c> {
    fn render_glyph(&mut self, c: char) -> anyhow::Result<Glyph> {
-      self.face.set_pixel_sizes(0, self.size_store.size)?;
-      self.face.load_char(c as usize, LoadFlag::RENDER)?;
-      let glyph_size = {
-         let glyph = self.face.glyph();
-         (glyph.bitmap().width(), glyph.bitmap().rows())
-      };
+      let size = self.size_store.size as f32;
+      let mut scaler = self.scale_context.builder(self.swash_font).size(size).hint(true).build();
+
+      let glyph_id = self.swash_font.charmap().map(c);
+      let image = Render::new(&[
+         Source::ColorOutline(0),
+         Source::ColorBitmap(StrikeWith::BestFit),
+         Source::Outline,
+      ])
+      .format(Format::Alpha)
+      .render(&mut scaler, glyph_id)
+      .unwrap(); // TODO: handle None value later
```

Comment:
So... what about this TODO'd `None` case? When can it happen? If it doesn't happen like, ever, maybe we should provide a better error message here by using `expect`?

Reply:
Yeah, in those cases you usually look at the source code. And skimming through it I can see that this can happen for a variety of reasons, essentially whenever the glyph cannot be rendered. IMO this case should be handled gracefully on our side and not render the glyph at all in case this happens.

Reply:
I think an empty glyph would work for the time being. The technically correct way would be to render tofu (the `.notdef` glyph) but we can introduce that sometime else.

Reply:
I think 0x0 pixels would work since we use a texture packer. Doing this will return the UV coordinates for whatever's at the packer's cursor, and will not take up any actual space. The texture may come out as garbage but that's fine in case of rendering errors.

This would probably not work if we had one texture per glyph, but the texture packer enables us to do crazy things like this.